include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -32,6 +32,69 @@ namespace LinearAlgebra
 
   namespace TpetraWrappers
   {
+    namespace internal
+    {
+      template <typename VectorType>
+      typename VectorType::value_type *
+      begin(VectorType &V)
+      {
+        return V.begin();
+      }
+
+      template <typename VectorType>
+      const typename VectorType::value_type *
+      begin(const VectorType &V)
+      {
+        return V.begin();
+      }
+
+      template <typename VectorType>
+      typename VectorType::value_type *
+      end(VectorType &V)
+      {
+        return V.end();
+      }
+
+      template <typename VectorType>
+      const typename VectorType::value_type *
+      end(const VectorType &V)
+      {
+        return V.end();
+      }
+
+      template <typename Number, typename MemorySpace>
+      Number *
+      begin(LinearAlgebra::TpetraWrappers::Vector<Number, MemorySpace> &V)
+      {
+        return V.trilinos_vector().getDataNonConst().get();
+      }
+
+      template <typename Number, typename MemorySpace>
+      const Number *
+      begin(const LinearAlgebra::TpetraWrappers::Vector<Number, MemorySpace> &V)
+      {
+        return V.trilinos_vector().getData().get();
+      }
+
+      template <typename Number, typename MemorySpace>
+      Number *
+      end(LinearAlgebra::TpetraWrappers::Vector<Number, MemorySpace> &V)
+      {
+        return V.trilinos_vector().getDataNonConst().get() +
+               V.trilinos_vector().getLocalLength();
+      }
+
+      template <typename Number, typename MemorySpace>
+      const Number *
+      end(const LinearAlgebra::TpetraWrappers::Vector<Number, MemorySpace> &V)
+      {
+        return V.trilinos_vector().getData().get() +
+               V.trilinos_vector().getLocalLength();
+      }
```

Comment:
Instead of providing these overloads here, why not just add `begin()` and `end()` functions to the `LinearAlgebra::TpetraWrappers::Vector` class? Then you won't need any of these functions here.

Reply:
Sure, I will change this!
But do all other vector classes In deal.II implement `begin()` and `end()` ?

I saw, at least `dealii::vector<Number>`  has a `begin()` and `end()` method implemented.

Reply:
Good question. I don't actually know whether all vectors have `begin()` and `end()`, but even if they did it's not clear what semantics that would follow: Only the locally owned? locally stored? all elements?

Perhaps what you have is better. In fact, perhaps it would be better to call these functions `begin_locally_owned()` and `end_locally_owned()` to make this clear.